sssd_test_framework/hosts/base.py

```diff
@@ -264,3 +265,36 @@ def distro_minor(self) -> int:
         if not self._os_release:
             self._distro_information()
         return self._distro_minor
+
+    def get_package_version(self, package: str = "sssd", raise_on_error: bool = True) -> tuple:
+        """
+        Parse package version and return it as a tuple:
+         major, minor, patch, release
+        :param package: package name
+        :param raise_on_error: raise exeption when package is missing
+        :return: major, minor, patch, release
+        :rtype: tuple
+        :raises OSError: If package is missing or version could not be parsed.
+        """
+        rpm = self.conn.run("test -f /usr/bin/rpm", raise_on_error=False)
+        dpkg = self.conn.run("test -f /usr/bin/dpkg-query", raise_on_error=False)
+        if rpm.rc == 0:
+            ver = self.conn.run(f'rpm -q {package} --queryformat "%{{VERSION}}-%{{RELEASE}}"').stdout
+        elif dpkg.rc != 0:
+            ver = self.conn.run(f"dpkg-query -f '${{Version}}' -W {package}").stdout
+        else:
+            if raise_on_error:
+                raise OSError(f"Package {package} not found!")
+            return 0, 0, 0, "unknown"
+
+        v_match = re.match(r"([0-9]+)\.([0-9]+)\.([0-9]+)[.-~](.*)", ver, re.IGNORECASE)
```

Comment:
I did try packaging initially but it did not work on sssd versioning, it just fails like this.  

packaging.version.InvalidVersion: Invalid version: '2.10.0~beta2-3'

Reply:
Understood.  I took a quick glance at trying to modify VERSION_PATTERN but, couldn't get it to match either.

I did get a modified regex to consider that matches a wider variety of patterns:
```
    v_match = re.match(r"([0-9]+)(?:\.)?([0-9]+)?(?:\.)?([0-9]+)?(?:~)?([a-z0-9]+)?[\.-]?(.*)?",
                       ver, re.IGNORECASE)
```

Unfortunately, I wasn't able to make it any shorter and still match the same way.  I'm sure there's a better regex to use but, this one seemed to work for the following cases:
```
1 =  1 None None None 
1.2 =  1 2 None None 
1.2.3 =  1 2 3 None 
1.2.3.4 =  1 2 3 None 4
1.2.3-4 =  1 2 3 None 4
1.2.3~beta.4 =  1 2 3 beta 4
1.2.3~beta-4 =  1 2 3 beta 4
```

This uses 5 groups for {major}.{minor}.{patch}~{pre-release-string}-{release} where basically everything after {major} is optional.

Also, is 0~beta2 what you want/expected for patch?  I did the above so that beta2 is pulled into it's own group/variable.

Reply:
Hi @spoore1,
I took Your regex as it is better than the original one and added one more part (update).
I did thought that I need only major minor and patch in most of the cases as features should not be
introduced in releases, that is why I did not focus on pre-release, release or update.